### PR TITLE
MRG: Refactor mlab object out of examples

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
#### Reference issue
Follow-up of #5822

#### What does this implement/fix?
I noticed the example scripts still use heavily the object `mlab` from Mayavi so this PR isolates those calls  behind the 3d backend abstraction layer without exposing the private class `_Renderer`.

### CODE REVIEW

File: mne/viz/backends/_pyvista.py [hunk 1]
```diff
@@ -332,6 +336,7 @@ def _get_view_to_display_matrix(size):
 
 
 def _close_all():
+    # XXX This is not implemented yet
```
Comment (agramfort) [inline mne/viz/backends/_pyvista.py]: raise NotImplementedError

File: mne/viz/backends/_pyvista.py [hunk 2]
```diff
@@ -253,6 +253,10 @@ def show(self):
     def close(self):
         self.plotter.close()
 
+    def title(self, text):
+        # XXX This is not implemented yet
```
Comment (agramfort) [inline mne/viz/backends/_pyvista.py]: NotImplementedError

File: mne/viz/_3d.py [hunk 3]
```diff
@@ -1122,8 +1128,23 @@ def plot_alignment(info=None, trans=None, subject=None, subjects_dir=None,
...
+                                                (0., 0., 0.)),
+                            distance=view.get('distance', 0.6))
+    else:
+        raise TypeError('view must be an instance of `dict` or `None`: '
+                        '{} was given'.format(type(view)))
+    if isinstance(title, str):
+        renderer.title(title)
+    elif title is not None:
+        raise TypeError('title must be an instance of `str` or `None`: '
+                        '{} was given'.format(type(title)))
```
Comment (agramfort) [inline mne/viz/_3d.py]: this should be put in a separate private function. If you decide to make the function public

mne.viz.set_camera(renderer or mlab.figure or scene, view)

you don't need to add any parameter for view to the public API.

thoughts someone?

File: mne/viz/_3d.py [hunk 4]
```diff
@@ -1122,8 +1128,23 @@ def plot_alignment(info=None, trans=None, subject=None, subjects_dir=None,
...
-                        focalpoint=(0., 0., 0.), distance=0.6)
+    if view is None:
+        renderer.set_camera(azimuth=90, elevation=90,
+                            focalpoint=(0., 0., 0.), distance=0.6)
+    elif isinstance(view, dict):
+        renderer.set_camera(azimuth=view.get('azimuth', 90),
+                            elevation=view.get('elevation', 90),
+                            focalpoint=view.get('focalpoint',
+                                                (0., 0., 0.)),
+                            distance=view.get('distance', 0.6))
```
Comment (massich) [inline mne/viz/_3d.py]: I like the 	`view.get('key', or_default)`. But If you do this, I would do something like:

```py
_check_value(view, (dict, None))
view = dict() if view is None else view
render.set_cammera(azimuth=view.get('azimuth', 90),
                            elevation=view.get('elevation', 90),
                            focalpoint=view.get('focalpoint',
                                                (0., 0., 0.)),
                            distance=view.get('distance', 0.6)))
```

and I even think that if you put the defaults in `set_cammera` and allow a `*kwargs` something like this should work:

```py
_check_value(view, (dict, None))
view = dict() if view is None else view
render.set_cammera(**view)
```

File: mne/viz/_3d.py (hunk 3 above)
Comment (massich) [reply mne/viz/_3d.py]: maybe a `kwargs` as extra parameters?

File: mne/viz/_3d.py [hunk 5]
```diff
@@ -2545,9 +2568,18 @@ def snapshot_brain_montage(fig, montage, hide_sensors=True):
     else:
         raise TypeError('montage must be an instance of `DigMontage`, `Info`,'
                         ' or `dict`')
+    if view is None:
+        view = dict()
+    if not isinstance(view, dict):
+        raise TypeError('view must be an instance of `dict` or None: '
+                        '{} was given'.format(type(view)))
```
Comment (massich) [inline mne/viz/_3d.py]: I would use the check

File: mne/viz/_3d.py (hunk 4 above)
Comment (massich) [reply mne/viz/_3d.py]: yes it works!! 

```py
def set_camera(azimuth=90, elevation=90, focalpoint=(0., 0., 0.),
               distance=0.6, **dump_parameters):
    print('azimuth :', azimuth,
          ', elevation :', elevation,
          ', focalpoint :', focalpoint,
          ', distance :', distance)


set_camera(**dict())
set_camera(**dict(azimuth='az', focalpoint='focal'))
set_camera(**dict(azimuth='az', not_here='not_here'))
```

Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: I like it, this is very compact and it's straightforward to pass the parameters that way.

Comment (massich) [reply mne/viz/_3d.py]: its a way of saying: I have this dict of paramameters use what you can.

Notice that you need `**dump_parameters` otherwise the 3rd call would give you an error. I would log/warn/error if `dump_parameters` is not empty dict

Comment (massich) [reply mne/viz/_3d.py]: or maybe is a good idea to not have `dump_parameters` like this if you pass something that it's not supposed to be passed it would crash. Up to you

File: mne/viz/_3d.py (hunk 3 above)
Comment (agramfort) [reply mne/viz/_3d.py]: please no kwargs in public functions if we can

File: mne/io/egi/egi.py [hunk 6]
```diff
@@ -24,8 +24,8 @@ def _read_header(fid):
     if version > 6 & ~np.bitwise_and(version, 6):
         version = version.byteswap().astype(np.uint32)
     else:
-        ValueError('Watchout. This does not seem to be a simple '
-                   'binary EGI file.')
+        raise ValueError('Watchout. This does not seem to be a simple '
+                         'binary EGI file.')
```
Comment (agramfort) [inline mne/io/egi/egi.py]: nice catch but this should be done is a different PR

Comment (GuillaumeFavelier) [reply mne/io/egi/egi.py]: I wasn't sure, I'll open one right now

Comment (GuillaumeFavelier) [reply mne/io/egi/egi.py]: By any chance, do you know why it has been reported by [LGTM](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-ed53039d31b3d79a6e49b8288ed75eefb81e3126) only now?

**Comment by agramfort** [comment]
LGTM gets better with time :)

>

File: mne/viz/_3d.py [hunk 7]
```diff
@@ -2559,6 +2559,88 @@ def snapshot_brain_montage(fig, montage, hide_sensors=True):
     return proj.xy, im
 
 
+def plot_sensors_connectivity(raw, picks, ch_names, con):
+    """Visualize the sensor connectivity in 3D.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        A Raw object containing FIF data.
```
Comment (larsoner) [inline mne/viz/_3d.py]: This should be `info`

File: mne/viz/_3d.py [hunk 8]
```diff
@@ -2559,6 +2559,88 @@ def snapshot_brain_montage(fig, montage, hide_sensors=True):
...
+    picks: array of int
+        Indices of selected channels.
+    ch_names: list of str
+        The names of the channels.
```
Comment (larsoner) [inline mne/viz/_3d.py]: No need for this, just use `info['ch_names']`

File: mne/viz/_3d.py [hunk 9]
```diff
@@ -2559,6 +2559,88 @@ def snapshot_brain_montage(fig, montage, hide_sensors=True):
...
 
+def plot_sensors_connectivity(raw, picks, ch_names, con):
+    """Visualize the sensor connectivity in 3D.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        A Raw object containing FIF data.
+    picks: array of int
+        Indices of selected channels.
```
Comment (larsoner) [inline mne/viz/_3d.py]: Move this after `con` and make the default `None` with `%(picks_data_good)s` or so

Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: When I tried to replace the docstring of `picks` by `%(picks_good_data)s` I got the following error:
```sh
E           AssertionError: 
E           mne.viz._3d.plot_sensors_connectivity %(picks_good_data)s != picks
```
How does it work?

**Comment by agramfort** [comment]
to use %(picks_good_data)s you need to use the verbose or fill_doc decorator

>

**Comment by GuillaumeFavelier** [comment]
A quick TODO list:

- [ ] find a good set of parameters in `test_plot_sensors_connectivity`
- [ ] add a `scalar_bar` function to the `_Renderer` API (used in `plot_sensors_connectivity`)
- [ ] add more options to `set_3d_title` if really necessary (look `plot_meg_sensors.py` example)

**Comment by GuillaumeFavelier** [comment]
Thanks for the help @agramfort , I'll update the TODO list!

**Comment by GuillaumeFavelier** [comment]
Once the TODO list is done, only `plot_decimate_head_surface.py` and `plot_read_inverse.py` remains.

**Comment by larsoner** [comment]
I'm not sure there is much value in `plot_decimate_head_surface.py`, maybe we should just get rid of it.

`plot_read_inverse` if it plots a source space can just use `plot_alignment`.

**Comment by GuillaumeFavelier** [comment]
Thanks for the feedback, I updated the TODO list accordingly.

**Comment by GuillaumeFavelier** [comment]
With your latest changes, @agramfort , we should be free of any `mlab` call in `examples` right?

**Comment by larsoner** [comment]
Nothing in `examples` but `tutorials` is another story:
```
larsoner@bunk:~/python/mne-python$ git grep ".*mayavi.*mlab"
...
tutorials/misc/plot_ecog.py:from mayavi import mlab
tutorials/misc/plot_sensor_locations.py:from mayavi import mlab
tutorials/sample-datasets/plot_brainstorm_phantom_elekta.py:from mayavi import mlab
tutorials/sample-datasets/plot_phantom_4DBTi.py:from mayavi import mlab
tutorials/simulation/plot_dics.py:from mayavi import mlab
tutorials/simulation/plot_point_spread.py:from mayavi import mlab
tutorials/source-modeling/plot_dipole_orientations.py:from mayavi import mlab
tutorials/source-modeling/plot_forward.py:from mayavi import mlab  # noqa
tutorials/source-modeling/plot_source_alignment.py:from mayavi import mlab
```

**Comment by GuillaumeFavelier** [comment]
I would prefer if we move step by step, the focus of this PR is mainly on `examples`. What do you say if we do a follow-up PR just for this?

**Comment by larsoner** [comment]
Fine by me

**Comment by GuillaumeFavelier** [comment]
I think I found a good compromise for `set_3d_title` (new `size` parameter) so with this the TODO list is done and I think the PR is ready for reviews/merge if the CIs come back happy.

I shall add that with the modifications of the `_Renderer` API here, some new features are not available yet in the `PyVista` backend. A follow-up PR might address this.

**Comment by larsoner** [comment]
> I shall add that with the modifications of the _Renderer API here, some new features are not available yet in the PyVista backend. A follow-up PR might address this.

Can you update the table so we can keep track?

**Comment by GuillaumeFavelier** [comment]
I already did for the new `plot_sensors_connectivity()` function but actually we have no table for the `_Renderer` "contract", what do you think is a good place for this?

**Comment by larsoner** [comment]
Ahh right it's not public. Just make sure it's documented somewhere so we don't forget, like a GitHub issue or code comment.

**Comment by GuillaumeFavelier** [comment]
To be more precise, the following functions of `mne.viz.backends._pyvista._Renderer` are not implemented yet:
* `tube()`
* `text3d()`
* `scalarbar()`
* `_set_3d_view()`
* `_set_3d_title()`

File: examples/visualization/plot_meg_sensors.py [hunk 10]
```diff
@@ -45,12 +43,11 @@
...
-    text.property.font_size = 40
-    mlab.draw(fig)
+    set_3d_title(figure=fig, title=system)
+    # text.x_position = 0.5
+    # text.y_position = 0.95
+    # text.property.vertical_justification = 'top'
+    # text.property.justification = 'center'
+    # text.actor.text_scale_mode = 'none'
+    # text.property.bold = True
+    # text.property.font_size = 40
```
Comment (agramfort) [inline examples/visualization/plot_meg_sensors.py]: @GuillaumeFavelier this needs to be cleaned up before merging

**Comment by agramfort** [comment]
@larsoner merge if happy

**Comment by massich** [comment]
+1

**Comment by agramfort** [comment]
@massich feel free to merge if diff lgty
